src/artifact-uploader.cpp

```diff
@@ -0,0 +1,258 @@
+
+    // all upload* methods are asynchronous, non-blocking, take ownership
+    // of data passed to them
+    void uploadBackground(const timestamp_t& timestamp, PayloadAuPtr payload);
+    void uploadObjectStream(const ObjectStream& stream, PayloadAuPtr payload);
+    void uploadFlipbook(const Flipbook& flipbook, PayloadAuPtr payload);
+    void uploadEvent(const timestamp_t& timestamp, move_ref<Events> data);
+
+private:
+    UploadQueuePtr uploadQueue_;
+    OutputControllerPtr outputController_;
+    UploadTaskQueuerPtr uploadTaskQueuer_;
+};
+
+ArtifactUploader::ArtifactUploader()
+    : pImpl_(new Impl())
+{
+
+}
+
+Status ArtifactUploader::init(const ArtifactUploader::Configuration& cfg,
+                              ArtifactUploader::ClientConfigCallback* configCallback)
+{
+    return pImpl_->init(cfg, configCallback);
+}
+
+void ArtifactUploader::uploadBackground(const timestamp_t& timestamp, PayloadAuPtr payload)
+{
+    pImpl_->uploadBackground(timestamp, payload);
+}
+
+void ArtifactUploader::uploadObjectStream(const ObjectStream& stream, PayloadAuPtr payload)
+{
+    pImpl_->uploadObjectStream(stream, payload);
+}
+
+void ArtifactUploader::uploadFlipbook(const Flipbook& flipbook, PayloadAuPtr payload)
+{
+    pImpl_->uploadFlipbook(flipbook, payload);
+}
+
+void ArtifactUploader::uploadEvent(const timestamp_t& timestamp, move_ref<Events> events)
+{
+    pImpl_->uploadEvent(timestamp, events);
+}
+
+ArtifactUploader::Impl::~Impl()
+{
+    if (uploadTaskQueuer_)
+        uploadTaskQueuer_->finalizeUpload();
+
+    if (outputController_)
+        outputController_->stop();
+}
+
+Status ArtifactUploader::Impl::init(const ArtifactUploader::Configuration& cfg,
+                                    ArtifactUploader::ClientConfigCallback* configCallback)
+{
+    if (cfg.maxQueueSizeMB <= 1e-5)
+    {
+        LOG(ERROR) << "Invalid maxQueueSizeMB value " << cfg.maxQueueSizeMB;
+        return makeError();
+    }
+
+    if (cfg.warnQueueSizeMB <= 1e-5)
+    {
+        LOG(ERROR) << "Invalid warnQueueSizeMB value " << cfg.warnQueueSizeMB;
+    }
+
+    uploadQueue_ = boost::make_shared<UploadQueue>(cfg.maxQueueSizeMB * 10e6, cfg.warnQueueSizeMB * 10e6);
+
+    PrismConnectServicePtr connect(new PrismConnectService());
+    PrismConnectService::Configuration serviceConfig;
+
+    serviceConfig.apiRoot = cfg.apiRoot;
+    serviceConfig.apiToken = cfg.apiToken;
+    serviceConfig.cameraName = cfg.cameraName;
+
+    int result = connect->init(serviceConfig);
+
+    if (result)
+    {
+        LOG(ERROR) << "Failed to init Prism Connect service, result = " << result;
+        return makeError();
+    }
+
+    outputController_ = boost::make_shared<OutputController>(
+                OutputController::Configuration(
+                    uploadQueue_,
+                    boost::make_shared<ArtifactUploadHelper>(connect),
+                    cfg.timeoutToCompleteUploadSec));
+
+    if (!cfg.queueType.compare(kStrSimple))
```

Comment:
I suggest moving this validation to the begin of the function. We don't need to create other stuff if queueType is invalid.
Also, uploadQueue seems to be created earlier than it's used and can be unnecessary if connect service init failed.